r1cs/gadgets/crypto/src/crh/poseidon/mod.rs

```diff
@@ -121,50 +122,48 @@ impl<ConstraintF: PrimeField + MulShortAssign<ConstraintF>, P: PoseidonParameter
 
             // Perform the matrix mix
             Self::matrix_mix (cs.ns(|| format!("poseidon_mix_matrix_partial_round_{}", _i)), state)?;
+        }
 
-            // Add the round constants to the state vector
+        // Second full rounds
+        for _i in 0..P::R_F {
+
+            // Add the round constants
             for d in state.iter_mut() {
                 let rc = P::ROUND_CST[round_cst_idx];
-                (*d).add_constant_in_place(cs.ns(|| format!("add_constant_2_{}", round_cst_idx)), &rc)?;
+                (*d).add_constant_in_place(cs.ns(|| format!("add_constant_3_{}", round_cst_idx)), &rc)?;
                 round_cst_idx += 1;
             }
 
-        }
-
-        // Second full rounds
-        // Process only to R_F -1 iterations. The last iteration does not contain a matrix mix
-        for _i in 0..(P::R_F-1) {
-
             // Apply the S-BOX to each of the elements of the state vector
             for (j, d) in state.iter_mut().enumerate() {
                 Self::mod_inv_sbox(cs.ns(||format!("mod_inv_S-Box_3_{}_{}",_i, j)), d)?;
             }
 
             // Perform the matrix mix
             Self::matrix_mix (cs.ns(|| format!("poseidon_mix_matrix_second_full_round_{}", _i)), state)?;
-
-            // Add the round constants
-            for d in state.iter_mut() {
-                //let rc = MNT4753Fr::from_str(ROUND_CST[round_cst_idx]).map_err(|_| ()).unwrap();
-                let rc = P::ROUND_CST[round_cst_idx];
-                (*d).add_constant_in_place(cs.ns(|| format!("add_constant_3_{}", round_cst_idx)), &rc)?;
-                round_cst_idx += 1;
-            }
         }
 
-        // Last full round does not perform the matrix_mix
-        {
-            // Apply the S-BOX to each of the elements of the state vector
-            for (j, d) in state.iter_mut().enumerate() {
-                Self::mod_inv_sbox(cs.ns(|| format!("mod_inv_S-Box_4_{}_{}", P::R_F-1, j)), d)?;
-            }
+        Ok(())
+    }
+
+    // Function that does the dot product for the mix matrix
+    fn dot_prod<CS: ConstraintSystem<ConstraintF>>(
```

Comment:
We should comment that this version propagates linear combinations, and hence is not meant to be used for proof systems which are sensitive on the R1CS density.